services/payments/tests/unit/test_services_socketio.py

```diff
@@ -0,0 +1,262 @@
+        },
+    )
+
+
+@pytest.fixture
+def user_primary_group_id(faker: Faker) -> GroupID:
+    return parse_obj_as(GroupID, faker.pyint())
+
+
+@pytest.fixture
+async def socketio_server(app: FastAPI) -> AsyncIterable[AsyncServer]:
+    # Same configuration as simcore_service_webserver/socketio/server.py
+    settings: RabbitSettings = get_rabbitmq_settings(app)
+    server_manager = AsyncAioPikaManager(url=settings.dsn)
+
+    server = AsyncServer(
+        async_mode="aiohttp",
+        engineio_logger=True,
+        client_manager=server_manager,
+    )
+
+    yield server
+
+    await cleanup_socketio_async_pubsub_manager(server_manager)
+
+
+@pytest.fixture
+def socketio_server_events(
+    socketio_server: AsyncServer,
+    mocker: MockerFixture,
+    user_primary_group_id: GroupID,
+) -> dict[str, AsyncMock]:
+
+    user_room_name = f"{user_primary_group_id}"
+
+    # handlers
+    async def connect(sid: str, environ):
+        print("connecting", sid)
+        await socketio_server.enter_room(sid, user_room_name)
+
+    async def on_check(sid, data):
+        print("check", sid, data)
+
+    async def on_payment(sid, data):
+        print("payment", sid, parse_obj_as(PaymentTransaction, data))
+
+    async def disconnect(sid: str):
+        print("disconnecting", sid)
+        await socketio_server.leave_room(sid, user_room_name)
+
+    # spies
+    spy_connect = mocker.AsyncMock(wraps=connect)
+    socketio_server.on("connect", spy_connect)
+
+    spy_on_payment = mocker.AsyncMock(wraps=on_payment)
+    socketio_server.on(SOCKET_IO_PAYMENT_COMPLETED_EVENT, spy_on_payment)
+
+    spy_on_check = mocker.AsyncMock(wraps=on_check)
+    socketio_server.on("check", spy_on_check)
+
+    spy_disconnect = mocker.AsyncMock(wraps=disconnect)
+    socketio_server.on("disconnect", spy_disconnect)
+
+    return {
+        connect.__name__: spy_connect,
+        disconnect.__name__: spy_disconnect,
+        on_check.__name__: spy_on_check,
+        on_payment.__name__: spy_on_payment,
+    }
+
+
+@pytest.fixture
+async def web_server(
+    socketio_server: AsyncServer, aiohttp_unused_port: Callable
+) -> AsyncIterator[URL]:
+    """
+    this emulates the webserver setup: socketio server with
+    an aiopika manager that attaches an aiohttp web app
+    """
+    aiohttp_app = web.Application()
+    socketio_server.attach(aiohttp_app)
+
+    async def _lifespan(
+        server: TestServer, started: asyncio.Event, teardown: asyncio.Event
+    ):
+        # NOTE: this is necessary to avoid blocking comms between client and this server
+        await server.start_server()
+        started.set()  # notifies started
+        await teardown.wait()  # keeps test0server until needs to close
+        await server.close()
+
+    setup = asyncio.Event()
+    teardown = asyncio.Event()
+
+    server = TestServer(aiohttp_app, port=aiohttp_unused_port())
+    t = asyncio.create_task(_lifespan(server, setup, teardown), name="server-lifespan")
```

Comment:
@pcrespov sorry, I didn't manage to review before the PR got in. But I really like this and I think it would be very useful to have this somewhere in one of the packages. I guess one could do this with any of the test servers we are using, right?